File: Ocis.Client/sdk/ts/Response.ts
import { Union } from "./fable_modules/fable-library-ts.4.25.0/Types.js";
import { union_type, string_type, TypeInfo } from "./fable_modules/fable-library-ts.4.25.0/Reflection.js";
import { ParseResult$1_$union, DeserializeResponse } from "./Protocol.js";
import { uint8 } from "./fable_modules/fable-library-ts.4.25.0/Int32.js";
import { ResponsePacket } from "./Ocis.Server/ProtocolSpec.js";
import { value as value_2, Option, defaultArg } from "./fable_modules/fable-library-ts.4.25.0/Option.js";

export type ClientResult$1_$union<T> = 
    | ClientResult$1<T, 0>
    | ClientResult$1<T, 1>
    | ClientResult$1<T, 2>

export type ClientResult$1_$cases<T> = {
    0: ["Success", [T]],
    1: ["NotFound", []],
    2: ["Error", [string]]
}

export function ClientResult$1_Success<T>(Item: T) {
    return new ClientResult$1<T, 0>(0, [Item]);
}

export function ClientResult$1_NotFound<T>() {
    return new ClientResult$1<T, 1>(1, []);
}

export function ClientResult$1_Error<T>(Item: string) {
    return new ClientResult$1<T, 2>(2, [Item]);
}

export class ClientResult$1<T, Tag extends keyof ClientResult$1_$cases<T>> extends Union<Tag, ClientResult$1_$cases<T>[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: ClientResult$1_$cases<T>[Tag][1]) {
        super();
    }
    cases() {
        return ["Success", "NotFound", "Error"];
    }
}

export function ClientResult$1_$reflection(gen0: TypeInfo): TypeInfo {
    return union_type("Ocis.Client.SDK.Response.ClientResult`1", [gen0], ClientResult$1, () => [[["Item", gen0]], [], [["Item", string_type]]]);
}

export function parseResponse(bytes: uint8[]): ParseResult$1_$union<ResponsePacket> {
    return DeserializeResponse(bytes);
}

export function toClientResult(parseResult: ParseResult$1_$union<ResponsePacket>): ClientResult$1_$union<void> {
    switch (parseResult.tag) {
        case /* ParseError */ 1:
            return ClientResult$1_Error<void>(parseResult.fields[0]);
        case /* InsufficientData */ 2:
            return ClientResult$1_Error<void>("Insufficient data");
        default: {
            const response: ResponsePacket = parseResult.fields[0];
            const matchValue: uint8 = response.StatusCode;
            switch (matchValue) {
                case 0:
                    return ClientResult$1_Success<void>(undefined);
                case 1:
                    return ClientResult$1_NotFound<void>();
                case 2:
                    return ClientResult$1_Error<void>(defaultArg(response.ErrorMessage, "Unknown error"));
                default:
                    return ClientResult$1_Error<void>("Invalid status code");
            }
        }
    }
}

export function toClientResultValue(parseResult: ParseResult$1_$union<ResponsePacket>): ClientResult$1_$union<uint8[]> {
    switch (parseResult.tag) {
        case /* ParseError */ 1:
            return ClientResult$1_Error<uint8[]>(parseResult.fields[0]);
        case /* InsufficientData */ 2:
            return ClientResult$1_Error<uint8[]>("Insufficient data");
        default: {
            const response: ResponsePacket = parseResult.fields[0];
            const matchValue: uint8 = response.StatusCode;
            switch (matchValue) {
                case 0: {
                    const matchValue_1: Option<uint8[]> = response.Value;
                    if (matchValue_1 == null) {
                        return ClientResult$1_Error<uint8[]>("Success response missing value");
                    }
                    else {
                        return ClientResult$1_Success<uint8[]>(value_2(matchValue_1));
                    }
                }
                case 1:
                    return ClientResult$1_NotFound<uint8[]>();
                case 2:
                    return ClientResult$1_Error<uint8[]>(defaultArg(response.ErrorMessage, "Unknown error"));
                default:
                    return ClientResult$1_Error<uint8[]>("Invalid status code");
            }
        }
    }
}

File: Ocis.Client/sdk/py/response.py
from __future__ import annotations
from typing import (Any, Generic, TypeVar)
from .Ocis_Server.protocol_spec import ResponsePacket
from .fable_modules.fable_library.option import default_arg
from .fable_modules.fable_library.reflection import (TypeInfo, string_type, union_type, enum_type)
from .fable_modules.fable_library.types import (Array, Union, uint8)
from .protocol import (DeserializeResponse, ParseResult_1)

_T = TypeVar("_T")

def _expr2(gen0: TypeInfo) -> TypeInfo:
    return union_type("Ocis.Client.SDK.Response.ClientResult`1", [gen0], ClientResult_1, lambda: [[("Item", gen0)], [], [("Item", string_type)]])


class ClientResult_1(Union, Generic[_T]):
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["Success", "NotFound", "Error"]


ClientResult_1_reflection = _expr2

def parse_response(bytes: bytearray) -> ParseResult_1[ResponsePacket]:
    return DeserializeResponse(bytes)


def to_client_result(parse_result: ParseResult_1[ResponsePacket]) -> ClientResult_1[None]:
    if parse_result.tag == 1:
        return ClientResult_1(2, parse_result.fields[0])

    elif parse_result.tag == 2:
        return ClientResult_1(2, "Insufficient data")

    else: 
        response: ResponsePacket = parse_result.fields[0]
        match_value: enum_type("Ocis.Server.ProtocolSpec.StatusCode", uint8, [("Success", 0.0), ("NotFound", 1.0), ("Error", 2.0)]) = response.StatusCode
        if match_value == uint8(0):
            return ClientResult_1(0, None)

        elif match_value == uint8(1):
            return ClientResult_1(1)

        elif match_value == uint8(2):
            return ClientResult_1(2, default_arg(response.ErrorMessage, "Unknown error"))

        else: 
            return ClientResult_1(2, "Invalid status code")




def to_client_result_value(parse_result: ParseResult_1[ResponsePacket]) -> ClientResult_1[bytearray]:
    if parse_result.tag == 1:
        return ClientResult_1(2, parse_result.fields[0])

    elif parse_result.tag == 2:
        return ClientResult_1(2, "Insufficient data")

    else: 
        response: ResponsePacket = parse_result.fields[0]
        match_value: enum_type("Ocis.Server.ProtocolSpec.StatusCode", uint8, [("Success", 0.0), ("NotFound", 1.0), ("Error", 2.0)]) = response.StatusCode
        if match_value == uint8(0):
            match_value_1: bytearray | None = response.Value
            if match_value_1 is None:
                return ClientResult_1(2, "Success response missing value")

            else: 
                return ClientResult_1(0, match_value_1)


        elif match_value == uint8(1):
            return ClientResult_1(1)

        elif match_value == uint8(2):
            return ClientResult_1(2, default_arg(response.ErrorMessage, "Unknown error"))

        else: 
            return ClientResult_1(2, "Invalid status code")




__all__ = ["ClientResult_1_reflection", "parse_response", "to_client_result", "to_client_result_value"]

